bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/discovery/ThingDiscoveryServiceTest.java

```diff
@@ -0,0 +1,246 @@
+/**
+ * Copyright (c) 2010-2023 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.boschshc.internal.discovery;
+
+import static org.hamcrest.CoreMatchers.*;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+import java.util.ArrayList;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.openhab.binding.boschshc.internal.devices.BoschSHCBindingConstants;
+import org.openhab.binding.boschshc.internal.devices.bridge.BridgeHandler;
+import org.openhab.binding.boschshc.internal.devices.bridge.dto.Device;
+import org.openhab.binding.boschshc.internal.devices.bridge.dto.Room;
+import org.openhab.core.config.discovery.DiscoveryResult;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.ThingUID;
+
+/**
+ * ThingDiscoveryService Tester.
+ *
+ * @author Gerd Zanker - Initial contribution
+ */
+@NonNullByDefault
+public class ThingDiscoveryServiceTest extends ThingDiscoveryService {
+
+    private int thingDiscoveredCalls = 0;
+    @Nullable
+    private DiscoveryResult lastDiscoveryResult = null;
+
+    @Override
+    protected void thingDiscovered(DiscoveryResult discoveryResult) {
+        thingDiscoveredCalls++;
+        lastDiscoveryResult = discoveryResult;
+        super.thingDiscovered(discoveryResult);
+    }
+
+    @Nullable
+    private ThingDiscoveryServiceTest fixture;
+
+    @Nullable
+    private BridgeHandler bridgeHandler;
+
+    @BeforeEach
+    void beforeEach() {
+        fixture = new ThingDiscoveryServiceTest();
```

Comment:
instead of instantiating the test class here you should instantiate an instance of the class under test (ThingDiscoveryService). Then you can also remove "extends ThingDiscoveryService"

Reply:
Ok, I will give it a try. I wasn't aware that the functions under test can still be called by an "unrelated" class.

Reply:
Hi @david-pace, can you please help here with your Java/mockito experience?

The ThingDiscoveryService calls the ``thingDiscovered`` function of its base class for each discovered SHC thing.
I want to intersept the calls during the test and evaluate the argument.

With a derived class, like the current ThingDiscoveryServiceTest, I'm able to do this. But with the ThingDiscoveryService 
 class under test I fail, because thingDiscovered is a protected function and mockito's spy can't support in this case.

Is it better to create a new ThingDiscoveryService derived class to use this "spy" class in the test code as the fixture instance? Do you know other possibilities?

Reply:
I am not aware of a way to achieve this with the standard Mockito toolset. It might be possible with PowerMockito, but this is a bit too risky/unstable in my experience. I think it is also not provided in the openHAB development setup.

I found another solution though: `thingDiscovered` will look for discovery listeners and will forward the result to them. So you can simply register a discovery listener in your `@BeforeEach` method. This listener can either be a custom implementation or also a mock. If it is a mock you can conveniently use Mockito's `verify()` methods to check how often it was called. And you can use argument captors to intercept arguments you are interested in. The `@BeforeEach` method then looks like this:

```
@BeforeEach
void beforeEach() {
    fixture = new ThingDiscoveryService();
    ...        
    DiscoveryListener discoveryListener = mock(DiscoveryListener.class);
    fixture.addDiscoveryListener(discoveryListener);
}
```

Look at `SmartBulbHandlerTest` to see an example for the combination of `verify()` and argument captors.